### PR TITLE
Parametrized SHOW STATS

### PR DESCRIPTION
**Pull request by fiedukow**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 1]
```diff
@@ -489,18 +486,19 @@ private static String getFunctionType(SqlFunction function)
...
+//
+//            List<String> statisticsNames = findUniqueStatisticsNames(tableStatistics);
+//
+//            List<String> resultColumnNames = buildColumnsNames(statisticsNames);
+//            SelectItem[] selectItems = buildSelectItems(resultColumnNames);
+//
+//            Map<ColumnHandle, String> columnNames = getStatisticsColumnNames(tableStatistics, node, table);
+//            List<Expression> resultRows = buildStatisticsRows(tableStatistics, columnNames, statisticsNames);
+//
+//            return simpleQuery(selectList(selectItems), aliased(new Values(resultRows), "table_stats_for_" + table, resultColumnNames));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: undo?

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 2]
```diff
@@ -663,7 +664,65 @@ public Node visitShowColumns(SqlBaseParser.ShowColumnsContext context)
...
+    {
+        // The following properties of SELECT subquery are required:
+        //  - only one relation in FROM
+        //  - only plain columns in projection
+        //  - only plain columns and constants in WHERE
+        //  - no group by
+        //  - no having
+        //  - no set quantifier
+        List<SqlBaseParser.RelationContext> relations = context.querySpecification().relation();
+        check(relations.size() != 0, "There must be exactly one relation in SHOW STATS SELECT clause", context.querySpecification());
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:681]: this should probably raise some semantic exception instead of illegal state

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 3]
```diff
@@ -1645,19 +1651,63 @@ public void testExists()
...
+                                                    new QualifiedNameReference(QualifiedName.of("field")),
+                                                    new LongLiteral("0"))
+                                            )
+                                    )));
+        }
+    }
+
+    private ShowStats createShowStats(QualifiedName name, List<SelectItem> selects, Optional<Expression> where)
+    {
+        return new ShowStats(new QuerySpecification(
```
Comment (kokosing) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: use `simpleQuery` instead of calling constructor

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/QuerySpecification.java [hunk 4]
```diff
@@ -56,7 +56,7 @@ public QuerySpecification(
         this(Optional.of(location), select, from, where, groupBy, having, orderBy, limit);
     }
 
-    private QuerySpecification(
+    public QuerySpecification(
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/QuerySpecification.java]: neh... why is so?

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 5]
```diff
@@ -483,22 +501,125 @@ private static String getFunctionType(SqlFunction function)
             throw new IllegalArgumentException("Unsupported function kind: " + kind);
         }
 
+        private static class NodesFinder extends PlanVisitor<Void, Void>
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: put it at the bottom of the file. 

Use `PlanNodeSearcher`

**Review by kokosing — COMMENTED** [review]
approach is ok

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java (hunk 2 above)
Comment (fiedukow) [reply presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:681]: It throws ParsingException as far as I know.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 5 above)
Comment (fiedukow) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: Thanks for pointer!

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java (hunk 3 above)
Comment (fiedukow) [reply presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: This would require to transit ShowStats from usage of QuerySpecification to Query, though it might be a good idea.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java [hunk 6]
```diff
@@ -124,7 +124,7 @@ public String getGraphvizPlan(Session session, Statement statement, Type planTyp
         throw new IllegalArgumentException("Unhandled plan type: " + planType);
     }
 
-    private Plan getLogicalPlan(Session session, Statement statement, List<Expression> parameters)
+    public Plan getLogicalPlan(Session session, Statement statement, List<Expression> parameters)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java:127]: extract as separate commit

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowStats.java [hunk 7]
```diff
@@ -13,30 +13,48 @@
...
 
-    private ShowStats(Optional<NodeLocation> location, QualifiedName table)
+    public ShowStats(Optional<NodeLocation> location, Query query)
     {
         super(location);
-        this.table = table;
+        this.query = query;
+    }
+
+    private static Query createFakeSpecification(Optional<NodeLocation> location, QualifiedName name)
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowStats.java]: what is that?

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 8]
```diff
@@ -489,55 +485,11 @@ private static String getFunctionType(SqlFunction function)
...
-
-        private static List<String> findUniqueStatisticsNames(TableStatistics tableStatistics)
-        {
-            TreeSet<String> statisticsKeys = new TreeSet<>();
-            statisticsKeys.addAll(tableStatistics.getTableStatistics().keySet());
-            for (ColumnStatistics columnStats : tableStatistics.getColumnStatistics().values()) {
-                statisticsKeys.addAll(columnStats.getStatistics().keySet());
-            }
-            return unmodifiableList(new ArrayList(statisticsKeys));
+            ShowStatsRewriter rewriter = new ShowStatsRewriter(metadata, session, parameters, queryExplainer.get());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: make it a separate StatementRewrite

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 9]
```diff
@@ -663,7 +664,67 @@ public Node visitShowColumns(SqlBaseParser.ShowColumnsContext context)
...
+        check(context.querySpecification().having == null, "HAVING is not supported in SHOW STATS SELECT clause", context.querySpecification().having);
+        check(context.querySpecification().setQuantifier() == null, "ALL/DISTINCT are not supported by SHOW STATS SELECT clause", context.querySpecification().setQuantifier());
+
+        QuerySpecification specification = (QuerySpecification) visitQuerySpecification(context.querySpecification());
+        int selectId = 0;
+        for (SelectItem selectItem : specification.getSelect().getSelectItems()) {
+            if (selectItem instanceof AllColumns) {
+                continue;
+            }
+            check(selectItem instanceof SingleColumn, "Only SingleColumn and AllColumn selects are supported by SHOW STATS SELECT clause", context.querySpecification().selectItem(selectId));
```
Comment (losipiuk) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: change this to `Only * and column references supported ...`

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 10]
```diff
@@ -663,7 +664,67 @@ public Node visitShowColumns(SqlBaseParser.ShowColumnsContext context)
...
+
+            SingleColumn columnSelect = (SingleColumn) selectItem;
+            check(columnSelect.getExpression() instanceof QualifiedNameReference, "Only FieldReference is allowed on SELECT list of SHOW STATS SELECT clause", context.querySpecification().selectItem(selectId));
```
Comment (losipiuk) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: ditto. Use same error message

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 11]
```diff
@@ -663,7 +664,67 @@ public Node visitShowColumns(SqlBaseParser.ShowColumnsContext context)
...
+
+        return new ShowStats(Optional.of(getLocation(context)), query);
+    }
+
+    public static final List<Class<? extends Expression>> ALLOWED_SHOW_STATS_WHERE_EXPRESSION_TYPES = ImmutableList.of(
+            Literal.class, QualifiedNameReference.class, ComparisonExpression.class, LogicalBinaryExpression.class, NotExpression.class);
+
+    void validateShowStatsWhereExpression(Expression expression, ParserRuleContext context)
+    {
+        check(ALLOWED_SHOW_STATS_WHERE_EXPRESSION_TYPES.stream().anyMatch(clazz -> clazz.isInstance(expression)), "Only Literals, FieldReferences, Comparisons, And, Ors and Not expression are allowed in WHERE of SHOW STATS SELECT clause", context);
```
Comment (losipiuk) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: `Only literals, column references, comparators and logical operators are allowed ...`

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 12]
```diff
@@ -487,13 +493,13 @@ public void testValues()
     public void testPrecedenceAndAssociativity()
             throws Exception
     {
-        assertExpression("1 AND 2 OR 3", new LogicalBinaryExpression(LogicalBinaryExpression.Type.OR,
```
Comment (losipiuk) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: Add some negative testing. Just a bit.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+        for (ColumnStatistics columnStats : tableStatistics.getColumnStatistics().values()) {
+            statisticsKeys.addAll(columnStats.getStatistics().keySet());
+        }
+        return unmodifiableList(new ArrayList(statisticsKeys));
+    }
+
+    private static Optional<TableScanNode> findTableScanNode(Plan plan)
+    {
+        List<TableScanNode> scanNodes = PlanNodeSearcher.searchFrom(plan.getRoot()).where(TableScanNode.class::isInstance).findAll();
+        checkState(scanNodes.size() <= 1, "There must be max of 1 table scan nodes in SHOW STATS SELECT clause");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java]: static import
I would add `getOnlyElement` to `PlanNodeSearcher` (like https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Iterables.html#getOnlyElement-java.lang.Iterable-T-)

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java [hunk 14]
```diff
@@ -38,7 +38,7 @@ public Select(NodeLocation location, boolean distinct, List<SelectItem> selectIt
         this(Optional.of(location), distinct, selectItems);
     }
 
-    private Select(Optional<NodeLocation> location, boolean distinct, List<SelectItem> selectItems)
+    public Select(Optional<NodeLocation> location, boolean distinct, List<SelectItem> selectItems)
```
Comment (kokosing) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/Select.java]: why public?

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 15]
```diff
@@ -503,7 +509,7 @@ public void testPrecedenceAndAssociativity()
                 new NotExpression(new LongLiteral("1")),
                 new LongLiteral("2")));
 
-        assertExpression("NOT 1 OR 2", new LogicalBinaryExpression(LogicalBinaryExpression.Type.OR,
+        assertExpression("NOT 1 OR 2", new LogicalBinaryExpression(OR,
```
Comment (kokosing) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: extract to separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+        if (tableLayouts.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(tableLayouts.get(0).getLayout().getHandle());
+    }
+
+    private Node createEmptyStatsTable(QualifiedName tableName)
+    {
+        return simpleQuery(selectList(unaliasedName("column_name")), aliased(new Values(ImmutableList.of()), "table_stats_for_" + tableName, ImmutableList.of("column_name")));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java]: As we discussed. IMO this method should be called `createStatsForEmptyLayout` and return table filled with zeros.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 17]
```diff
@@ -1645,19 +1651,60 @@ public void testExists()
...
+            assertStatement(format("SHOW STATS ON %s", qualifiedName), new ShowStats(qualifiedName));
         }
     }
 
     @Test
+    public void testShowStatsParametrized()
+    {
+        final String[] tableNames = {"t", "s.t", "c.s.t"};
+
+        for (String fullName : tableNames) {
```
Comment (kokosing) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:1670]: is this loop needed? I don't see much value in it

**Review by kokosing — APPROVED** [review]
minor comments

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+        this.queryExplainer = requireNonNull(queryExplainer, "queryExplainer is null");
+    }
+
+    public Node rewrite(ShowStats node)
+    {
+        checkArgument(node.getQuery().getQueryBody() instanceof QuerySpecification);
+        QuerySpecification specification = (QuerySpecification) node.getQuery().getQueryBody();
+
+        checkArgument(specification.getFrom().isPresent());
+        Table table = (Table) specification.getFrom().get();
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java]: some checkstate that this actually will be Table?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java (hunk 6 above)
Comment (fiedukow) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java:127]: As far as I know Presto has exactly oposite policy on extracting such changes.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+    {
+        List<TableScanNode> scanNodes = PlanNodeSearcher.searchFrom(plan.getRoot()).where(TableScanNode.class::isInstance).findAll();
+        checkState(scanNodes.size() <= 1, "There must be max of 1 table scan nodes in SHOW STATS SELECT clause");
+        return getOnlyElement(scanNodes.stream().map(Optional::of).collect(toImmutableList()), Optional.empty());
+    }
+
+    private static Expression findFilterExpression(Plan plan)
+    {
+        List<FilterNode> filterNodes = PlanNodeSearcher.searchFrom(plan.getRoot()).where(TableScanNode.class::isInstance).findAll();
+        checkState(filterNodes.size() <= 1, "There must be max of 1 table filter nodes in SHOW STATS SELECT clause");
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java]: `there must be at most one filter node ...`

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java (hunk 13 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java]: `there must be at most one ....`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowStats.java (hunk 7 above)
Comment (fiedukow) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowStats.java]: This should be called createFakeQuery and it's used to handle two syntaxes using only one rewrite. We create fake query for `SHOW STATS qualifiedName` case as there is no query there.
Creating this query may be moved to AstBuilder if you like that more.

**Comment by losipiuk** [comment]
Makes sense. I would opt for returnign zeros for no-layout situation though.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java (hunk 15 above)
Comment (fiedukow) [reply presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: changes removed

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 8 above)
Comment (fiedukow) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: I don't like that idea, but here you go.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java (hunk 17 above)
Comment (fiedukow) [reply presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java:1670]: It covers cases we managed to do wrong when we first started implementing that with @ArturGajowy

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java (hunk 16 above)
Comment (fiedukow) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewriter.java]: We can do that if we decide it's better answer. I dont see any value in that.

**Comment by fiedukow** [comment]
merged
